x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/netty4/SecurityNetty4ServerTransport.java

```diff
@@ -99,6 +123,80 @@ protected void initChannel(final Channel ch) throws Exception {
 
     }
 
+    @Override
+    protected Netty4MessageInboundHandler getNetty4MessageInboundHandler(String name) {
+        if (remoteClusterPortEnabled == false || REMOTE_CLUSTER_PROFILE.equals(name) == false || apiKeyService == null) {
+            return super.getNetty4MessageInboundHandler(name);
+        }
+
+        final ThreadPool threadPool = getThreadPool();
+        final Transport.RequestHandlers requestHandlers = getRequestHandlers();
+        final InboundAggregator aggregator = new InboundAggregator(
+            getInflightBreaker(),
+            requestHandlers::getHandler,
+            ignoreDeserializationErrors()
+        ) {
+            @Override
+            public void headerReceived(Header header) {
+                super.headerReceived(header);
+
+                if (header.isRequest() == false) {
+                    throw new IllegalArgumentException("Remote cluster server cannot process response");
+                }
+
+                if (header.isHandshake()) {
+                    return;
+                }
+
+                final String credentials = header.getHeaders().v1().get(CROSS_CLUSTER_ACCESS_CREDENTIALS_HEADER_KEY);
+                final ApiKeyService.ApiKeyCredentials apiKeyCredentials;
+                if (credentials != null) {
+                    apiKeyCredentials = ApiKeyService.getCredentialsFromHeader(credentials, ApiKey.Type.CROSS_CLUSTER);
+                } else {
+                    apiKeyCredentials = null;
+                }
+
+                if (apiKeyCredentials == null) {
+                    logger.warn("No credential is found");
+                    shortCircuit(new ElasticsearchSecurityException("No credential is found", RestStatus.UNAUTHORIZED));
+                    return;
+                }
+
+                final ThreadContext threadContext = threadPool.getThreadContext();
+                final var contextPreservingActionListener = new ContextPreservingActionListener<AuthenticationResult<User>>(
+                    threadContext.wrapRestorable(threadContext.newStoredContext()),
+                    ActionListener.wrap(authenticationResult -> {
+                        if (false == authenticationResult.isAuthenticated()) {
+                            logger.warn("authentication failed [{}]", authenticationResult.getMessage());
+                            shortCircuit(
+                                authenticationResult.getException() != null
+                                    ? authenticationResult.getException()
+                                    : new ElasticsearchSecurityException(authenticationResult.getMessage(), RestStatus.UNAUTHORIZED)
+                            );
+                        } else {
+                            logger.info("authentication successful");
+                        }
+                    }, this::shortCircuit)
```

Comment:
This probably needs to be something like `channel().eventLoop().submit(() -> this::shortCircuit)`.
The assumption everywhere is that this is not multithreaded code.

Reply:
I could be wrong. But I don't think using event loop is necessary in this case. If we agree with the general approach, my plan is to have a named subclass for the Aggregator where we can use an AtomicReference for the exception to short circuit. Thread safety will be ensured by it. Also, we don't really care exactly what exception is set by which thread as long as there is one to signify short circuit.